### ISSUE TITLE
Inconsistent behavior of client.Users.get

### ISSUE THREAD

**Issue by btubbs** [synapsepay/SynapsePayRest-Python#3]
If you call client.Users.get(), without passing a user_id argument, you get a list of users.

If you call it with a user_id argument, you get back just that user's payload.

If you then use that same client object, and call .get() with no argument again, you will not get a list of users, as you might expect, but the same user that you got back in the last call.

Similar, nastier bugs could arise in production from the way this user_id state is maintained on the client object.  If you're working with multiple users at once, it would be very easy to do something you didn't intend.

I see two possible fixes, both of which assume that the client object stops maintaining/using a user_id attribute.

1. Require explicitly passing in a user_id for all calls like update(), attach_file(), etc.
2. Instead of Users.get() passing back users as just dictionaries, it could pass back instances of a new User object.  That new object would have a reference to the client, and also a copy of the user_id.  It could then provide methods like update, attach_file, etc, that would not require passing in a user_id.

**Comment by sankaet** [comment]
@btubbs 

`If you then use that same client object, and call .get() with no argument again, you will not get a list of users, as you might expect, but the same user that you got back in the last call.`

This seems like a bug. We will patch this here shortly.

Regarding 

`Similar, nastier bugs could arise in production from the way this user_id state is maintained on the client object. If you're working with multiple users at once, it would be very easy to do something you didn't intend.`

Each client object is intended to be used for one user at a time. We manage the oauth key for the user and also the user id. When working with multiple users, it is advised to initialize a new client object instead of overwriting an existing one. Otherwise all the user calls would not only require you to provide `user_ids`, but also the `oauth_key` and `fingerprint` for each user.

**Comment by folz** [comment]
> Each client object is intended to be used for one user at a time[....] When working with multiple users, it is advised to initialize a new client object instead of overwriting an existing one.

This confused me because I assumed .get (and the client in general) would be stateless / side-effect-less. I am only using one Client object, and it seems like keeping track of multiple client objects for multiple synapse IDs is cumbersome and unpythonic. Do you recommend something like this?

```
clients = {
    '12345': client.Client().get('12345'),
    '67890': client.Client().get('67890')
}
```

Agreed with "nastier bugs" too - since I don't think the client stores state in a thread-safe manner, I could see a situation where two simultaneous requests override each others' state. E.g. with gunicorn (I believe) this could be a nasty bug to track down

Request 1: client.Users.get('12345')
Request 2: client.Users.get('67890')
Request 1: client.Users.add_doc(...) # is created using synapse ID 67890

Let me know if I'm missing something on my end or if I can answer any questions.

**Comment by sankaet** [comment]
@folz client is stateless unless specified (in context to the user).

But The idea behind every client object is to manage all of the [initialization](http://docs.synapsepay.com/docs/api-initialization) states all the way to the user level.

That is actually the most important aspect of the libraries, it manages `_id`, `oauth_key`, `fingerprint`, `ip`, `client_id` & `client_secret`. If it just managed `client_id` & `client_secret` you will have to specify the others in payloads every time.


Regarding your 2nd point:

```
Request 1: client.Users.get('12345')
Request 2: client.Users.get('67890')
Request 1: client.Users.add_doc(...) # is created using synapse ID 67890
```

This is exactly what you are not supposed to do, you are supposed to do this:

```
Request 1: client1.Users.get('12345')
Request 2: client2.Users.get('67890')
Request 1: client1.Users.add_doc(...) # is created using synapse ID 12345
Request 2: client1.Users.add_doc(...) # is created using synapse ID 67890
```

**Comment by btubbs** [comment]
You're obviously free to design the client library how you think is best, but I think the stateful behavior of this library is surprising, and likely to trip people up.

I understand that you want to help cut down on boilerplate by managing the auth stuff automagically.  That is good!  But I think it would be a lot less confusing to do that on a User() object, and leave client stateless.  Otherwise you'll need to document the stateful nature of the client with giant flashing warnings, and still deal with a lot of people getting it wrong.

**Comment by sankaet** [comment]
@btubbs I think that is a good suggestion and something we can definitely work on

**Comment by sankaet** [comment]
This is how the new lib would function. I think this will make things easier.

```

synapse-python

pip install synapse-python

```

# Clients()

```

client = Clients(client_id='id-27ea1e51-b6bb-4cdd-9608-4a9e47f5feba',client_secret='secret-343fabc9-faca-4970-855b-51a88cf015da')

```

# Users(Clients)

```

users = Users.get(client=client,_id='56226c2786c2736898e7807c')

```

or, if you want to fetch a single user:

```

user = Users.get(client=client,_id='56226c2786c2736898e7807c',fingerprint='11c94ba6bad74d24a0158bc707f0fc19a86dc08f',ip='24.130.174.164')

```

If you want to create a new user:

```
user = Users.create(client=client,payload=playload)

```

Now once you have the user, you can update information:

```

user = user.update(payload=payload)

```


# Nodes(Users)

```

nodes = Nodes.get(user=user)

```

or, if you want to fetch a single node:

```

node = Nodes.get(user=user,_id='56226c2786c2736898e7807c')

```

If you want to create a new node:

```

node = Nodes.create(user=user,payload=payload)

```

Now once you have the node, you can update/delete node:

```

node = node.update(payload=payload)

```

or

```

node = node.delete(payload=payload)

```


# Transactions(Nodes)

```

transactions = Transactions.get(node=node)

```

or, if you want to fetch a single transaction:

```

transaction = Transactions.get(node=node,_id='56226c2786c2736898e7807c')

```

Now once you have the transaction, you can update/delete it:

```

transaction = transaction.update(payload=payload)

```

or

```

transaction = transaction.delete(payload=payload)

```

# Misc Functions


We will also add some misc functions to help around the APIs

```

padded_base_64 = url_to_base64(url='https://github.com/synapsepay/SynapsePayRest-Python/issues/3')

padded_base_64 = file_to_base64(path='/synapsepay/SynapsePayRest-Python/issues/3')

boolean_response = validate_hmac(hmac='56226c2786c2736898e7807c',client_secret='563db3fb86c27307d925871f',object_id='56226c2786c2736898e7807c')

```

**Comment by anthonysutardja** [comment]
++ for new lib spec.

Other suggestions:

- it would be nice to accept a byte stream as well in the misc functions
- python3 support

**Comment by sankaet** [comment]
@anthonysutardja thanks for the suggestions, we will add both

**Comment by stevula** [comment]
should be addressed by #5
